server/src/main/java/org/opensearch/common/cache/stats/CacheStatsHolder.java

```diff
@@ -0,0 +1,304 @@
+        // First try to increment without creating nodes
+        boolean didIncrement = internalIncrementHelper(dimensionValues, statsRoot, 0, adder, false);
+        // If we failed to increment, because nodes had to be created, obtain the lock and run again while creating nodes if needed
+        if (!didIncrement && createNodesIfAbsent) {
+            try {
+                lock.lock();
+                internalIncrementHelper(dimensionValues, statsRoot, 0, adder, true);
+            } finally {
+                lock.unlock();
+            }
+        }
+    }
+
+    /**
+     * Use the incrementer function to increment/decrement a value in the stats for a set of dimensions.
+     * If createNodesIfAbsent is true, and there is no stats for this set of dimensions, create one.
+     * Returns true if the increment was applied, false if not.
+     */
+    private boolean internalIncrementHelper(
+        List<String> dimensionValues,
+        Node node,
+        int depth, // Pass in the depth to avoid having to slice the list for each node.
+        Consumer<Node> adder,
+        boolean createNodesIfAbsent
+    ) {
+        if (depth == dimensionValues.size()) {
+            // This is the leaf node we are trying to reach
+            adder.accept(node);
+            return true;
+        }
+
+        Node child = node.getChild(dimensionValues.get(depth));
+        if (child == null) {
+            if (createNodesIfAbsent) {
+                boolean createMapInChild = depth < dimensionValues.size() - 1;
+                child = node.createChild(dimensionValues.get(depth), createMapInChild);
+            } else {
+                return false;
+            }
+        }
+        if (internalIncrementHelper(dimensionValues, child, depth + 1, adder, createNodesIfAbsent)) {
+            // Function returns true if the next node down was incremented
+            adder.accept(node);
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Produce an immutable version of these stats.
+     */
+    public ImmutableCacheStatsHolder getImmutableCacheStatsHolder() {
+        ImmutableCacheStatsHolder.Node snapshot = new ImmutableCacheStatsHolder.Node("", true, statsRoot.getImmutableStats());
+        // Traverse the tree and build a corresponding tree of MDCSDimensionNode, to pass to MultiDimensionCacheStats.
+        if (statsRoot.getChildren() != null) {
+            for (Node child : statsRoot.getChildren().values()) {
+                getImmutableCacheStatsHelper(child, snapshot);
+            }
+        }
+        return new ImmutableCacheStatsHolder(snapshot, dimensionNames);
+    }
+
+    private void getImmutableCacheStatsHelper(Node currentNodeInOriginalTree, ImmutableCacheStatsHolder.Node parentInNewTree) {
+        ImmutableCacheStatsHolder.Node newNode = createMatchingImmutableCacheStatsHolderNode(currentNodeInOriginalTree);
+        parentInNewTree.getChildren().put(newNode.getDimensionValue(), newNode);
```

Comment:
This is a call to `put` on the "immutable" node, which makes it not immutable.

Your better bet is to build the tree bottom-up, so that each immutable parent receives its (already-constructed) immutable children in its constructor.

You could add a method to the mutable `Node` to take a snapshot:

```
class Node {
  ImmutableCacheStatsHolder.Node snapshot() {
    Map<String, ImmutableCachStatsHolder.Node> snapshotChildren = new TreeMap<>();
    for (Node child : children.values()) {
      snapshotChildren.put(child.getDimensionValue(), child.snapshot());
    }
    if (snapshotChildren.isEmpty()) {
      snapshotChildren = Collections.emptyMap(); // replace empty with singleton
    } else {
      snapshotChildren = Collections.unmodifiableMap(snapshotChildren);
    }
    return new ImmutableCacheStatsHolder.Node(dimensionValue, snapshotChildren, getImmutableStats());
  }
}
```

Then the `getImmutableCacheStatsHolder()` method would just be:
```
public ImmutableCacheStatsHolder getImmutableCacheStatsHolder() {
  return new ImmutableCacheStatsHolder(statsRoot.snapshot(), dimensionNames);
```

And you could remove the two helper methods from this class.

Reply:
Good point - I like this approach, it's very clean.